File: electron/sqlite/queries/materials.js
/**
 * 材料与费用 SQLite 查询模块
 *
 * 实现 materials + expenses 两张表的 CRUD 操作。
 */
import log from 'electron-log';
import { tryGetSqlite, rowToCamel, toSqliteValue } from './helpers';
// ═══════════════════════════════════════════════════════════════════════════════
// 材料列映射
// ═══════════════════════════════════════════════════════════════════════════════
var MAT_COLUMNS = {
    id: 'id',
    projectId: 'project_id',
    name: 'name',
    category: 'category',
    unit: 'unit',
    quantity: 'quantity',
    price: 'price',
    createdAt: 'created_at',
};
var MAT_INSERT_COLS = ['project_id', 'name', 'category', 'unit', 'quantity', 'price', 'created_at'];
var MAT_INSERT_SQL = "INSERT INTO materials (".concat(MAT_INSERT_COLS.map(function (c) { return "\"".concat(c, "\""); }).join(', '), ") VALUES (").concat(MAT_INSERT_COLS.map(function () { return '?'; }).join(', '), ")");
// ═══════════════════════════════════════════════════════════════════════════════
// 材料读操作
// ═══════════════════════════════════════════════════════════════════════════════
/** 列出材料（可按项目过滤） */
export function listMaterials(projectId) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var sql = 'SELECT * FROM materials';
        var params = [];
        if (projectId) {
            sql += ' WHERE project_id = ?';
            params.push(projectId);
        }
        sql += ' ORDER BY created_at DESC';
        var rows = (_a = sqlite.prepare(sql)).all.apply(_a, params);
        return rows.map(rowToCamel);
    }
    catch (err) {
        log.error('[SQLite] materials.list error:', err);
        return null;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// 材料写操作
// ═══════════════════════════════════════════════════════════════════════════════
/** 创建材料 */
export function createMaterial(mat) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        sqlite.prepare(MAT_INSERT_SQL).run(toSqliteValue(mat.projectId), mat.name, toSqliteValue(mat.category || ''), toSqliteValue(mat.unit || ''), toSqliteValue(mat.quantity || 0), toSqliteValue(mat.price || 0), toSqliteValue(mat.createdAt || new Date().toISOString()));
        return true;
    }
    catch (err) {
        log.error('[SQLite] materials.create error:', err);
        return false;
    }
}
/** 更新材料 */
export function updateMaterial(id, changes) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var setClauses = [];
        var params = [];
        for (var _i = 0, _b = Object.entries(changes); _i < _b.length; _i++) {
            var _c = _b[_i], jsonKey = _c[0], value = _c[1];
            var col = MAT_COLUMNS[jsonKey];
            if (!col || col === 'id')
                continue;
            setClauses.push("\"".concat(col, "\" = ?"));
            params.push(toSqliteValue(value));
        }
        if (setClauses.length === 0)
            return true;
        params.push(id);
        var sql = "UPDATE materials SET ".concat(setClauses.join(', '), " WHERE id = ?");
        var result = (_a = sqlite.prepare(sql)).run.apply(_a, params);
        return result.changes > 0;
    }
    catch (err) {
        log.error('[SQLite] materials.update error:', err);
        return false;
    }
}
/** 删除材料 */
export function deleteMaterial(id) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var result = sqlite.prepare('DELETE FROM materials WHERE id = ?').run(id);
        return result.changes > 0;
    }
    catch (err) {
        log.error('[SQLite] materials.delete error:', err);
        return false;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// 费用列映射
// ═══════════════════════════════════════════════════════════════════════════════
var EXP_COLUMNS = {
    id: 'id',
    projectId: 'project_id',
    amount: 'amount',
    category: 'category',
    description: 'description',
    date: 'date',
    createdAt: 'created_at',
};
var EXP_INSERT_COLS = ['project_id', 'amount', 'category', 'description', 'date', 'created_at'];
var EXP_INSERT_SQL = "INSERT INTO expenses (".concat(EXP_INSERT_COLS.map(function (c) { return "\"".concat(c, "\""); }).join(', '), ") VALUES (").concat(EXP_INSERT_COLS.map(function () { return '?'; }).join(', '), ")");
// ═══════════════════════════════════════════════════════════════════════════════
// 费用读操作
// ═══════════════════════════════════════════════════════════════════════════════
/** 列出费用（可按项目过滤） */
export function listExpenses(projectId) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var sql = 'SELECT * FROM expenses';
        var params = [];
        if (projectId) {
            sql += ' WHERE project_id = ?';
            params.push(projectId);
        }
        sql += ' ORDER BY date DESC';
        var rows = (_a = sqlite.prepare(sql)).all.apply(_a, params);
        return rows.map(rowToCamel);
    }
    catch (err) {
        log.error('[SQLite] expenses.list error:', err);
        return null;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// 费用写操作
// ═══════════════════════════════════════════════════════════════════════════════
/** 创建费用 */
export function createExpense(exp) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        sqlite.prepare(EXP_INSERT_SQL).run(toSqliteValue(exp.projectId), toSqliteValue(exp.amount || 0), toSqliteValue(exp.category || ''), toSqliteValue(exp.description || ''), toSqliteValue(exp.date || ''), toSqliteValue(exp.createdAt || new Date().toISOString()));
        return true;
    }
    catch (err) {
        log.error('[SQLite] expenses.create error:', err);
        return false;
    }
}
/** 更新费用 */
export function updateExpense(id, changes) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var setClauses = [];
        var params = [];
        for (var _i = 0, _b = Object.entries(changes); _i < _b.length; _i++) {
            var _c = _b[_i], jsonKey = _c[0], value = _c[1];
            var col = EXP_COLUMNS[jsonKey];
            if (!col || col === 'id')
                continue;
            setClauses.push("\"".concat(col, "\" = ?"));
            params.push(toSqliteValue(value));
        }
        if (setClauses.length === 0)
            return true;
        params.push(id);
        var sql = "UPDATE expenses SET ".concat(setClauses.join(', '), " WHERE id = ?");
        var result = (_a = sqlite.prepare(sql)).run.apply(_a, params);
        return result.changes > 0;
    }
    catch (err) {
        log.error('[SQLite] expenses.update error:', err);
        return false;
    }
}
/** 删除费用 */
export function deleteExpense(id) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var result = sqlite.prepare('DELETE FROM expenses WHERE id = ?').run(id);
        return result.changes > 0;
    }
    catch (err) {
        log.error('[SQLite] expenses.delete error:', err);
        return false;
    }
}

File: src-tauri/src/commands/materials.rs
/**
 * 材料管理命令
 *
 * 对应 Electron 版本的 materials.ts（材料部分）
 * 表: materials(id, project_id, name, category, unit, quantity, price, created_at)
 */

use crate::db::AppState;
use crate::error::{AppError, AppResult};
use rusqlite::params;
use serde::{Deserialize, Serialize};
use tauri::{command, State};

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Material {
    pub id: i64,
    pub project_id: i64,
    pub name: String,
    pub category: Option<String>,
    pub unit: Option<String>,
    pub quantity: f64,
    pub price: f64,
    pub created_at: Option<String>,
    // JOIN 字段
    pub project_name: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NewMaterial {
    pub project_id: i64,
    pub name: String,
    pub category: Option<String>,
    pub unit: Option<String>,
    pub quantity: Option<f64>,
    pub price: Option<f64>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MaterialUpdate {
    pub name: Option<String>,
    pub category: Option<String>,
    pub unit: Option<String>,
    pub quantity: Option<f64>,
    pub price: Option<f64>,
}

/// 获取材料列表（可按项目过滤）
#[command]
pub fn get_materials(
    state: State<'_, AppState>,
    project_id: Option<i64>,
) -> AppResult<Vec<Material>> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let sql = if project_id.is_some() {
        "SELECT m.id, m.project_id, m.name, m.category, m.unit, m.quantity,
                m.price, m.created_at, p.name as project_name
         FROM materials m
         LEFT JOIN projects p ON m.project_id = p.id
         WHERE m.project_id = ?1
         ORDER BY m.created_at DESC"
    } else {
        "SELECT m.id, m.project_id, m.name, m.category, m.unit, m.quantity,
                m.price, m.created_at, p.name as project_name
         FROM materials m
         LEFT JOIN projects p ON m.project_id = p.id
         ORDER BY m.created_at DESC"
    };

    let mut stmt = db
        .prepare(sql)
        .map_err(|e| AppError::Database(e.to_string()))?;

    let materials = if let Some(pid) = project_id {
        stmt.query_map(params![pid], |row| {
            Ok(Material {
                id: row.get(0)?,
                project_id: row.get(1)?,
                name: row.get(2)?,
                category: row.get(3)?,
                unit: row.get(4)?,
                quantity: row.get(5)?,
                price: row.get(6)?,
                created_at: row.get(7)?,
                project_name: row.get(8)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?
    } else {
        stmt.query_map([], |row| {
            Ok(Material {
                id: row.get(0)?,
                project_id: row.get(1)?,
                name: row.get(2)?,
                category: row.get(3)?,
                unit: row.get(4)?,
                quantity: row.get(5)?,
                price: row.get(6)?,
                created_at: row.get(7)?,
                project_name: row.get(8)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?
    };

    Ok(materials)
}

/// 新增材料
#[command]
pub fn create_material(state: State<'_, AppState>, material: NewMaterial) -> AppResult<i64> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    db.execute(
        "INSERT INTO materials (project_id, name, category, unit, quantity, price)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6)",
        params![
            material.project_id,
            material.name,
            material.category,
            material.unit,
            material.quantity.unwrap_or(0.0),
            material.price.unwrap_or(0.0),
        ],
    )
    .map_err(|e| AppError::Database(e.to_string()))?;

    let id = db.last_insert_rowid();
    Ok(id)
}

/// 更新材料
#[command]
pub fn update_material(
    state: State<'_, AppState>,
    id: i64,
    updates: MaterialUpdate,
) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute(
            "UPDATE materials SET
             name = COALESCE(?1, name),
             category = COALESCE(?2, category),
             unit = COALESCE(?3, unit),
             quantity = COALESCE(?4, quantity),
             price = COALESCE(?5, price)
             WHERE id = ?6",
            params![
                updates.name,
                updates.category,
                updates.unit,
                updates.quantity,
                updates.price,
                id,
            ],
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("材料 {} 不存在", id)));
    }

    Ok(())
}

/// 删除材料
#[command]
pub fn delete_material(state: State<'_, AppState>, id: i64) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute("DELETE FROM materials WHERE id = ?1", params![id])
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("材料 {} 不存在", id)));
    }

    Ok(())
}
